### ISSUE TITLE
Library semantics

### ISSUE THREAD

**Issue by JosefWN** [shamblett/coap#75]
A couple of things that puzzled me as I've dug into the code. I also want to stress that I understand that this is originally a port of another project, so I'm sure it carries some legacy from there, and I don't mean to nag or be a PITA. I've just created this issue to spark some discussion regarding possibilities revolving around the future direction of the repo.

1. The request _is_ the client (part of it, anyway), as opposed to being something which is _processed_ by the client.

This is not only confusing semantically but might also pose issues if this library was to support non-blocking sends in the future. Imagine:
```dart
  final request = CoapRequest.newGet()..addUriPath('large');
  client.request = request;
  client.get().then((response) => print(response.payloadString));

  final request2 = CoapRequest.newGet()..addUriPath('hello');
  client.request = request2;
  client.get().then((response) => print(response.payloadString));

  print(client.request); // Two ongoing requests, which one is `client.request`?
``` 

Makes more sense to:
```dart
  final request = CoapRequest.newGet()..addUriPath('large');
  client.send(request).then((response) => print(response.payloadString));

  final request2 = CoapRequest.newGet()..addUriPath('hello');
  client.send(request2).then((response) => print(response.payloadString));
```
I know you can do both (synchronously), but should you be able to? In a larger piece of code, even with synchronous sends, with `client.send(request)` it's clear what's going on without any added context, whereas with `client.get()` it's much less so.

**Proposal**: Drop `client.request` and related methods.

2. Same thing with the endpoint, I would have imagined a 1:1 mapping between the client and a socket/endpoint. If you want to send requests to other endpoints you create other clients.

Confusing example now:
```dart
  final conf = CoapConfig();
  final uri = Uri(scheme: 'coap', host: 'coap.me', port: conf.defaultPort);
  final client = CoapClient(uri, conf);

  final request = CoapRequest.newGet()
    ..addUriPath('large')
    ..endpoint = notCoapDotMe;
  var response = await client.send(request);
```

The client URI says [coap.me](https://coap.me) but the request is sent somewhere else.

**Proposal**: Make `request.endpoint` a private field, or drop it altogether if the refactoring is feasible. The request still has a `destination` via its inheritance from `CoapMessage`, and `CoapRequest send()` could perhaps be moved to the client. Clients typically send requests, requests don't send themselves.

3. The request is a request and a response combined:

```dart
  CoapResponse? _currentResponse;
  CoapResponse? get response => _currentResponse;
```

4. The `threading` folder naming is a bit confusing. There is no threading going on there from what I can see. The executor is scheduling tasks on the main isolate, concurrency is not the same thing as parallellism.

**Proposal**: Rename the folder to `tasks` or `scheduling`?

**Comment by JKRhb** [comment]
Thank you for raising these points, @JosefWN! I also got the impression that it would be nice to rework the library in a number of aspects.

Regarding 1.: Maybe the new API could also look like this?

```dart
  final request = CoapRequest.newGet()..addUriPath('large');
  request.send().then((response) => print(response.payloadString));

  final request2 = CoapRequest.newGet()..addUriPath('hello');
  request2.send().then((response) => print(response.payloadString));
```

This might be a bit cleaner than first creating the request with the client and then passing it back to it to send it.

**Comment by JosefWN** [comment]
Hi @JKRhb!

That would shift the responsibility of sending the request from the client to the request itself. It's a matter of taste I suppose, but I would shift as much logic as possible to the client, and just consider the request something that encodes into valid CoAP packet(s).

https://pub.dev/packages/http seems like a popular package, could make sense to align with their top-level API to some extent since HTTP and CoAP are pretty similar protocols?

A shorthand could be:
```dart
var response = await client.get('hello');
```

So rather than doing away with the `get` method altogether as I initially proposed, it would be a shorthand for:
```dart
final request = CoapRequest.newGet()..addUriPath('hello');
var response = await client.send(request);
```

**Comment by JosefWN** [comment]
Speaking of library semantics, this issue also ties in with: https://github.com/shamblett/coap/issues/41

**Comment by JKRhb** [comment]
> That would shift the responsibility of sending the request from the client to the request itself. It's a matter of taste I suppose, but I would shift as much logic as possible to the client, and just consider the request something that encodes into valid CoAP packet(s).

Oh, you are right, it is certainly better to move everything to the client :)

> https://pub.dev/packages/http seems like a popular package, could make sense to align with their top-level API to some extent since HTTP and CoAP are pretty similar protocols?

I think that sounds like a good idea! Two cases that might require more attention here are probably observe and multicast. In the former case, I guess you could simply return a stream of responses?

```dart
// The possibility to distinguish between GET and FETCH would be needed here
final observeStream = client.observe('hello');
observeStream .listen((response) {
  ...
};
```

The case of multicast could probably be handled by letting the Response class implement the stream interface itself and then emitting events on each (following?) response? 

```dart
final multicastResponse = await client.get('coap://[ff02::fd]/hello');
multicastResponse.listen((response) {
  ...
};
```

I guess you could also add separate methods for multicast requests but it is probably desirable to have one unified API for both types of requests.

**Comment by JosefWN** [comment]
> Oh, you are right, it is certainly better to move everything to the client :)

Yeah it could be beneficial, since there are some things that span multiple requests (socket, tracking the message id sequence...).

```dart
final observeStream = client.observe('hello');
observeStream .listen((response) {
  ...
};
```

I think this makes sense. I'm no expert on multicast to be honest, since it's pretty different from the other request types/"verbs" it might be better to have a separate object, or perhaps a `CoapMulticastResponse` which could extend `CoapResponse` if they are similar enough.

Could also have optional named parameters in addition to path:
```dart
final getResponse = await client.get('hello', options: ...);
final postResponse = await client.post('hello', payload: ...);
```

... and/or support the full URI like you are suggesting, but then we might need to cache multiple sockets/endpoints over some time.

**Comment by shamblett** [comment]
OK, this is now merged into master, great work thanks guys. Do you want a new release here to baseline this or do you want to add the other few PR's to remove testserver example etc before we do this?

**Comment by JosefWN** [comment]
Great! 🚀

Perhaps we could:
- Include: https://github.com/shamblett/coap/pull/85, https://github.com/shamblett/coap/pull/82
-  Discuss and possibly implement this issue (also a breaking change): https://github.com/shamblett/coap/issues/78

Then release a 4.0.0?

**Comment by shamblett** [comment]
Sounds like a plan, PR's #85 and #82 merged.

**Comment by shamblett** [comment]
Package re released at version 4.0.0

**Comment by JosefWN** [comment]
Library size reduced by about 85%, performance improved by 5.5% averaged over 100 simple get requests when compared to v3.5.0. Yay! 😄

**Comment by shamblett** [comment]
Yes, this work has been well worth it.
